### PR TITLE
Move diagnostic from parser to later phase.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs [hunk 1]
```diff
@@ -2158,8 +2158,13 @@ public class Extensions
    public static void Main(){} 
 } 
 ";
-
-            ParseAndValidate(test, Diagnostic(ErrorCode.ERR_IdentifierExpectedKW, "this").WithArguments("", "this"));
+            CreateCompilationWithMscorlib(test).VerifyDiagnostics(
+                // (6,22): error CS0027: Keyword 'this' is not available in the current context
+                //    public Extensions(this int i) {} 
+                Diagnostic(ErrorCode.ERR_ThisInBadContext, "this").WithLocation(6, 22),
+                // (2,1): hidden CS8019: Unnecessary using directive.
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: Simplify the test by removing the unnecessary using?

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs [hunk 2]
```diff
@@ -2402,19 +2416,7 @@ static void M()
...
-                Diagnostic(ErrorCode.ERR_LbraceExpected, "this").WithLocation(6, 25),
-                // (6,25): error CS1002: ; expected
-                Diagnostic(ErrorCode.ERR_SemicolonExpected, "this").WithLocation(6, 25),
-                // (6,30): error CS1002: ; expected
-                Diagnostic(ErrorCode.ERR_SemicolonExpected, "object").WithLocation(6, 30),
-                // (6,38): error CS1002: ; expected
-                Diagnostic(ErrorCode.ERR_SemicolonExpected, ")").WithLocation(6, 38),
-                // (6,38): error CS1513: } expected
-                Diagnostic(ErrorCode.ERR_RbraceExpected, ")").WithLocation(6, 38));
+            CreateCompilationWithMscorlib(test).VerifyDiagnostics();
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: Why is there no diagnostic here? Delegate methods don't allow `this`.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs [hunk 3]
```diff
@@ -2379,15 +2387,21 @@ public void CS1041ERR_IdentifierExpectedKW03()
...
         [Fact, WorkItem(541347, "http://vstfdevdiv:8080/DevDiv2/DevDiv/_workitems/edit/541347")]
         public void CS1041ERR_IdentifierExpectedKW04()
         {
             var test = @"delegate void D(this object o);";
 
-            ParseAndValidate(test, Diagnostic(ErrorCode.ERR_IdentifierExpectedKW, "this").WithArguments("", "this"));
+            CreateCompilationWithMscorlib(test).VerifyDiagnostics(
+                // (1,17): error CS0027: Keyword 'this' is not available in the current context
+                // delegate void D(this object o);
+                Diagnostic(ErrorCode.ERR_ThisInBadContext, "this").WithLocation(1, 17));
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: I see the tests affected cover constructor, delegate, method, and property scenarios.
Could you add a similar test for user defined operator?

**Comment by CyrusNajmabadi** [comment]
This PR should be closed.  It was superceded.

File: src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs (hunk 3 above)
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Syntax/Parsing/ParserErrorMessageTests.cs]: sure!

**Pull request closed without merging**